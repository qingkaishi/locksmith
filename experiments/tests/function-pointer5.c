/*
 * There is an extra dereference in ok/function-pointer5.out compared to
 * out/function-pointer5.out.  I've left that in, because it's not a duplicate.
 * Hashconsing on guards got rid of duplicates and this one dereference.  The
 * soundness does not change, but when this scenario occurs (no instantiation
 * in a cycle) we might lose one dereference and only keep the one with the
 * longest call-path.
 * This is because adding one element to the path at the function-pointer
 * callsite, will give a guard with everything else being the same, and the
 * call path extended by one.  This will hide the preexisting guard that has a
 * sub-path for its callpath.  In all other cases it's fine to do so (and even
 * preferrable) but here we might miss a real dereference with an actually
 * shorter path.  Reporting the longer path is still correct but might not be
 * as readable.
 */

#include <pthread.h>
#include <stdio.h>

typedef void (*fptr)(void*);

int i;

void bar(void* arg) {
  fptr x = (fptr) arg;
  i++;//race
  x(arg);
}

fptr foo = &bar;

int main() {
  pthread_t t1;
  pthread_create(&t1, NULL, foo, foo);
  foo(foo);
}


