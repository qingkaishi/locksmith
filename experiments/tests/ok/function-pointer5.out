************************* STARTING *************************

shared: &i:function-pointer5.c:22

Warning: Possible data race: &i:function-pointer5.c:22 is not protected!
 references:
  dereference of &i:function-pointer5.c:22 at function-pointer5.c:26
    &i:function-pointer5.c:22
  locks acquired:
    <empty>
  in: main at function-pointer5.c:32  -> function-pointer5.c:35
  in: main at function-pointer5.c:32  -> function-pointer5.c:35 -> function-pointer5.c:27
  in: main at function-pointer5.c:32 
  in: main at function-pointer5.c:32  -> function-pointer5.c:27


*************************** DONE ***************************
